iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,215 @@
+        /// Creates an instance of this class. Provided for unit testing purposes only.
+        /// </summary>
+        protected FileUploadNotificationProcessorClient()
+        {
+        }
+
+        internal FileUploadNotificationProcessorClient(
+            string hostName,
+            IotHubConnectionProperties credentialProvider,
+            IotHubServiceClientOptions options)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _connection = new IotHubConnection(credentialProvider, options.UseWebSocketOnly, options);
+            FileNotificationReceiver = new AmqpFileNotificationReceiver(_connection);
+        }
+
+        /// <summary>
+        /// Gets the AmqpFileNotificationReceiver which receives notifications for file upload operations.
+        /// </summary>
+        internal AmqpFileNotificationReceiver FileNotificationReceiver;
+
+        /// <summary>
+        /// Open the connection and start receiving file upload notifications.
+        /// </summary>
+        /// <exception cref="IotHubCommunicationException">Thrown if the client encounters a transient retriable exception. </exception>
+        /// <exception cref="IotHubCommunicationException">Thrown when the operation has been canceled. The inner exception will be
+        /// <see cref="OperationCanceledException"/>.</exception>
+        /// <exception cref="SocketException">Thrown if a socket error occurs.</exception>
+        /// <exception cref="WebSocketException">Thrown if an error occurs when performing an operation on a WebSocket connection.</exception>
+        /// <exception cref="IOException">Thrown if an I/O error occurs.</exception>
+        /// <exception cref="IotHubException">Thrown if an error occurs when communicating with IoT hub service.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>true</c> then it is a transient exception.
+        /// If <see cref="IotHubException.IsTransient"/> is set to <c>false</c> then it is a non-transient exception.</exception>
+        public virtual async Task OpenAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening FileUploadNotificationProcessorClient", nameof(OpenAsync));
+
+            try
+            {
+                if (_fileNotificationProcessor == null)
+                {
+                    throw new Exception("Callback for message feedback {0} must be set before opening the connection.".FormatInvariant(_fileNotificationProcessor));
+                }
+                await FileNotificationReceiver.OpenAsync().ConfigureAwait(false);
+                ReceivingAmqpLink receivingAmqpLink = await FileNotificationReceiver.FaultTolerantReceivingLink.GetReceivingLinkAsync().ConfigureAwait(false);
+                receivingAmqpLink.RegisterMessageListener(OnNotificationMessageReceivedAsync);
+                receivingAmqpLink.Session.Connection.Closed += ConnectionClosed;
```

Comment:
We should also add our ```ConnectionClosed``` handler to these events as well:

```csharp
receivingAmqpLink.Session.Closed += ConnectionClosed;
receivingAmqpLink.Closed += ConnectionClosed;
```

For the customer's purposes, losing any of the link and/or session is basically equivalent to losing the connection/connection in that they won't be able to receive anything anymore.